Refactor clean_expr so that it returns an exprt and takes argument by copy

romainbrenguier opened the pull request:
This makes the information flow clearer, avoid errors where a subtype of
exprt could be passed as argument and its id mutated to something
inconsistent with this subtype, and avoids having to make explicit
copies in the places where the function is called.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [na] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
clean_expr returns exprt and takes argument by copy

This makes the information flow clearer, avoid errors where a subtype of
exprt could be passed as argument and its id mutated to something
inconsistent with this subtype, and avoids having to make explicit
copies in the places where the function is called.

tautschnig commented on src/goto-symex/symex_clean_expr.cpp:
```diff
@@ -213,8 +212,8 @@ void goto_symext::lift_lets(statet &state, exprt &rhs)
 
-void goto_symext::clean_expr(
-  exprt &expr,
+exprt goto_symext::clean_expr(
```
Could you please introduce a macro wrapping `__attribute__((warn_unused_result))` - I thought that @smowton had already done this, but now I can't find it. Otherwise changes of this kind result in bugs that are terrible to debug.

smowton replied:
It was on an abandoned branch of the let-expr work, but it's just a port of test-gen's `NODISCARD` macro so should be easily recreated

tautschnig reviewed (changes requested):
Looks good except we need to make sure we warn about unused return values.

romainbrenguier committed:
Add NODISCARD macro

This is used to mark functions for which the result should not be
ignored, and can be checked by some compilers.

romainbrenguier committed:
Mark clean_expr as NODISCARD

This will make some compilers check that the result of the function is
not ignored.

tautschnig commented:
@romainbrenguier clang-format and cpp-lint aren't quite happy yet.
